frame/cere-staking/src/lib.rs

```diff
@@ -0,0 +1,599 @@
+	}
+}
+
+/// Preference of what happens regarding to participating in storage network.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo, Default)]
+pub struct StoragePrefs {
+	foo: bool, // ToDo
+}
+
+/// Preference of what happens regarding to participating in CDN.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo, Default)]
+pub struct EdgePrefs {
+	foo: bool, // ToDo
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::{
+		pallet_prelude::*, sp_runtime::traits::StaticLookup, traits::LockableCurrency,
+		Blake2_128Concat,
+	};
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Currency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Number of eras that staked funds must remain bonded for.
+		#[pallet::constant]
+		type BondingDuration: Get<EraIndex>;
+	}
+
+	/// Map from all locked "stash" accounts to the controller account.
+	#[pallet::storage]
+	#[pallet::getter(fn bonded)]
+	pub type Bonded<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, T::AccountId>;
+
+	/// The minimum active bond to become and maintain the role of a storage network participant.
+	#[pallet::storage]
+	pub type MinStorageBond<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+	/// The minimum active bond to become and maintain the role of a CDN participant.
+	#[pallet::storage]
+	pub type MinEdgeBond<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+	/// Map from all (unlocked) "controller" accounts to the info regarding the staking.
+	#[pallet::storage]
+	#[pallet::getter(fn ledger)]
+	pub type Ledger<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, StakingLedger<T::AccountId, BalanceOf<T>>>;
+
+	/// The map from (wannabe) storage network participant stash key to the preferences of that
+	/// storage network participant.
+	#[pallet::storage]
+	#[pallet::getter(fn storages)]
+	pub type Storages<T: Config> =
+		CountedStorageMap<_, Twox64Concat, T::AccountId, StoragePrefs, ValueQuery>;
+
+	/// The map from (wannabe) CDN participant stash key to the preferences of that CDN participant.
+	#[pallet::storage]
+	#[pallet::getter(fn edges)]
+	pub type Edges<T: Config> =
+		CountedStorageMap<_, Twox64Concat, T::AccountId, EdgePrefs, ValueQuery>;
+
+	/// The current era index.
+	///
+	/// This is the latest planned era, depending on how the Session pallet queues the validator
+	/// set, it might be active or not.
+	#[pallet::storage]
+	#[pallet::getter(fn current_era)]
+	pub type CurrentEra<T> = StorageValue<_, EraIndex>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
```

Comment:
Let's have events on all state changes. For instance, we need to add event when `serve` and `store` executed.

Reply:
The original staking pallet does not emit an event from `validate` or `nominate` calls. Let's preserve this behavior? Other significant state updates are covered by existing events.

https://github.com/Cerebellum-Network/pos-network-node/blob/c026e5a11f3ced2fbaaca194c44e017b5e7e915a/frame/staking/src/pallet/mod.rs#L994
https://github.com/Cerebellum-Network/pos-network-node/blob/c026e5a11f3ced2fbaaca194c44e017b5e7e915a/frame/staking/src/pallet/mod.rs#L1034